block/fork.go

```diff
@@ -0,0 +1,266 @@
+//
+// This method checks two conditions to decide if a node should be stopped:
+// 1. If the next state height is greater than or equal to the rollapp's revision start height.
+// 2. If the block's app version (equivalent to revision) is less than the rollapp's revision
+func (m *Manager) shouldStopNode(rollapp *types.Rollapp, block *types.Block) bool {
+	revision := block.Header.Version.App
+	if m.State.NextHeight() >= rollapp.RevisionStartHeight && revision < rollapp.Revision {
+		m.logger.Info(
+			"Freezing node due to fork update",
+			"local_block_height",
+			m.State.Height(),
+			"rollapp_revision_start_height",
+			rollapp.RevisionStartHeight,
+			"local_revision",
+			revision,
+			"rollapp_revision",
+			rollapp.Revision,
+		)
+		return true
+	}
+
+	return false
+}
+
+// forkNeeded returns true if the fork file exists
+func (m *Manager) forkNeeded() (types.Instruction, bool) {
+	if instruction, err := types.LoadInstructionFromDisk(m.RootDir); err == nil {
+		return instruction, true
+	}
+
+	return types.Instruction{}, false
+}
+
+// handleSequencerForkTransition handles the sequencer fork transition
+func (m *Manager) handleSequencerForkTransition(instruction types.Instruction) {
+	consensusMsgs, err := m.prepareDRSUpgradeMessages(instruction.FaultyDRS)
+	if err != nil {
+		panic(fmt.Sprintf("prepare DRS upgrade messages: %v", err))
+	}
+	// Always bump the account sequences
+	consensusMsgs = append(consensusMsgs, &sequencers.MsgBumpAccountSequences{Authority: authtypes.NewModuleAddress("sequencers").String()})
+
+	err = m.handleForkBlockCreation(instruction, consensusMsgs)
+	if err != nil {
+		panic(fmt.Sprintf("validate existing blocks: %v", err))
+	}
+
+	if err := m.handleForkBatchSubmission(instruction.RevisionStartHeight); err != nil {
+		panic(fmt.Sprintf("ensure batch exists: %v", err))
+	}
+}
+
+// prepareDRSUpgradeMessages prepares consensus messages for DRS upgrades.
+// It performs version validation and generates the necessary upgrade messages for the sequencer.
+//
+// The function implements the following logic:
+//   - If no faulty DRS version is provided (faultyDRS is nil), returns no messages
+//   - Validates the current DRS version against the potentially faulty version
+//   - Generates an upgrade message with the current valid DRS version
+func (m *Manager) prepareDRSUpgradeMessages(faultyDRS []uint32) ([]proto.Message, error) {
+	if faultyDRS == nil {
+		return nil, nil
+	}
+
+	actualDRS, err := strconv.ParseUint(version.DrsVersion, 10, 32)
+	if err != nil {
+		return nil, fmt.Errorf("converting DRS version to int: %v", err)
+	}
+
+	for _, drs := range faultyDRS {
+		if drs == uint32(actualDRS) {
+			return nil, fmt.Errorf("running faulty DRS version %d", drs)
+		}
+	}
+
+	return []proto.Message{
+		&sequencers.MsgUpgradeDRS{
+			Authority:  authtypes.NewModuleAddress("sequencers").String(),
+			DrsVersion: uint64(actualDRS),
+		},
+	}, nil
```

Comment:
It looks like it's always upgrading DRS? We only want to do it if the last is faulty right